lib/pacer_test.go

```diff
@@ -0,0 +1,148 @@
+package vegeta
+
+import (
+	"math"
+	"testing"
+	"time"
+)
+
+func TestRatePacing(t *testing.T) {
+	t.Parallel()
+
+	for i, tt := range []struct {
+		freq    int
+		per     time.Duration
+		elapsed time.Duration
+		count   uint64
+		wait    time.Duration
+		stop    bool
+	}{
+		// NOTE: Rate sleeps before sending the first hit,
+		// rather than sending the first hit instantly!
+		// 1 hit/sec, 0 hits sent, 1s elapsed => 0s until next hit
+		// (time.Sleep will return immediately in this case)
+		{1, time.Second, time.Second, 0, 0, false},
+		// 1 hit/sec, 0 hits sent, 2s elapsed => -1s until next hit
+		// (time.Sleep will return immediately in this case)
+		{1, time.Second, 2 * time.Second, 0, -1 * time.Second, false},
+		// 1 hit/sec, 1 hit sent, 1s elapsed => 1s until next hit
+		{1, time.Second, time.Second, 1, time.Second, false},
+		// 1 hit/sec, 2 hits sent, 1s elapsed => 2s until next hit
+		{1, time.Second, time.Second, 2, 2 * time.Second, false},
+		// 1 hit/sec, 10 hits sent, 1s elapsed => 10s until next hit
+		{1, time.Second, time.Second, 10, 10 * time.Second, false},
+		// 1 hit/sec, 10 hits sent, 11s elapsed => 0s until next hit
+		{1, time.Second, 11 * time.Second, 10, 0, false},
+		// 2 hit/sec, 9 hits sent, 4.9s elapsed => 100ms until next hit
+		{2, time.Second, (49 * time.Second) / 10, 9, 100 * time.Millisecond, false},
+	} {
+		r := &Rate{Freq: tt.freq, Per: tt.per}
+
+		wait, stop := r.Pace(tt.elapsed, tt.count)
+		if have, want := wait, tt.wait; have != want {
+			t.Errorf("test case %d: %+v: wait=%v, want %v", i, r, have, want)
+		}
+		if have, want := stop, tt.stop; have != want {
+			t.Errorf("test case %d: %+v: stop=%v, want %v", i, r, have, want)
+		}
+	}
+}
+
+var quarterPeriods = map[string]float64{
+	"MeanUp":   MeanUp,
+	"Peak":     Peak,
+	"MeanDown": MeanDown,
+	"Trough":   Trough,
+}
+
+type sineTest struct {
+	p, m, a int
```

Comment:
Please give these more descriptive names and document how they're used. It's hard for me to understand otherwise.

Reply:
Sorry. I wrote these tests quickly, and I'm very lazy when it comes to testing :-(

Reply:
No problem, I understand! It's important to have the tests as maintainable as possible though.